src/azure/blobstore.rs

```diff
@@ -0,0 +1,303 @@
+// Copyright 2018 Benjamin Bader
+// Copyright 2016 Mozilla Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use azure::credentials::*;
+use base64;
+use crypto::digest::Digest;
+use crypto::hmac::Hmac;
+use crypto::mac::Mac;
+use crypto::md5::Md5;
+use crypto::sha2::Sha256;
+use futures::{Future, Stream};
+use hyper::{self, header, Method, Uri};
+use hyper::client::{Client, Request, HttpConnector};
+use hyper_tls::HttpsConnector;
+use std::fmt;
+use std::str::FromStr;
+use time;
+use tokio_core::reactor::Handle;
+
+use errors::*;
+
+const BLOB_API_VERSION: &str = "2017-04-17";
+
+fn hmac<D: Digest>(digest: D, data: &[u8], secret: &[u8]) -> Vec<u8> {
+    let mut hmac = Hmac::new(digest, secret);
+    hmac.input(data);
+    hmac.result().code().iter().map(|b| *b).collect::<Vec<u8>>()
+}
+
+fn signature(to_sign: &str, secret: &str) -> String {
+    let decoded_secret = base64::decode_config(secret.as_bytes(), base64::STANDARD).unwrap();
+    let sig = hmac(Sha256::new(), to_sign.as_bytes(), &decoded_secret);
+    base64::encode_config::<Vec<u8>>(&sig, base64::STANDARD)
+}
+
+fn md5(data: &[u8]) -> String {
+    let mut result: Vec<u8> = vec![0; 16]; // md5 digest is 16 bytes long.
+    let mut digest = Md5::new();
+    digest.input(data);
+    digest.result(&mut result);
+
+    base64::encode_config::<Vec<u8>>(&result, base64::STANDARD)
+}
+
+pub struct BlobContainer {
+    url: String,
+    client: Client<HttpsConnector<HttpConnector>>,
+}
+
+impl fmt::Display for BlobContainer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "BlobContainer(url={})", self.url)
+    }
+}
+
+impl BlobContainer {
+    pub fn new(base_url: &str, container_name: &Option<String>, handle: &Handle) -> Result<BlobContainer> {
+        let container_url = match container_name {
+            &Some(ref name) => format!("{}{}/", base_url, name), // base_url is assumed to end in a trailing slash
```

Comment:
Feels like it would be better to back this assumption up with a check in the code rather than leave it open for people to hit weird failures.